firestore/integration_test_internal/src/aggregate_query_snapshot_test.cc

```diff
@@ -0,0 +1,283 @@
+  EXPECT_TRUE(moved_snapshot_dest.is_valid());
+}
+
+TEST_F(AggregateQuerySnapshotTest, MoveAssignmentOperator) {
+  const Query& query = TestFirestore()->Collection("foo").Limit(10);
+  const AggregateQuery& aggregate_query = query.Count();
+
+  static const int COUNT = 3;
+  AggregateQuerySnapshot snapshot =
+      TestAggregateQuerySnapshot(aggregate_query, COUNT);
+
+  AggregateQuerySnapshot snapshot_move_dest = std::move(snapshot);
+
+  EXPECT_EQ(snapshot.count(), 0);
+  EXPECT_EQ(snapshot.query(), AggregateQuery());
+  EXPECT_FALSE(snapshot.is_valid());
+
+  EXPECT_EQ(snapshot_move_dest.count(), COUNT);
+  EXPECT_EQ(snapshot_move_dest.query(), aggregate_query);
+  EXPECT_TRUE(snapshot_move_dest.is_valid());
+}
+
+TEST_F(AggregateQuerySnapshotTest, Equality1) {
+  CollectionReference collection =
+      Collection({{"a", {{"k", FieldValue::String("a")}}},
+                  {"b", {{"k", FieldValue::String("b")}}},
+                  {"c", {{"k", FieldValue::String("c")}}}});
+  AggregateQuerySnapshot snapshot1 = ReadAggregate(collection.Limit(1).Count());
+  AggregateQuerySnapshot snapshot2 = ReadAggregate(collection.Limit(1).Count());
+
+  EXPECT_TRUE(snapshot1 == snapshot1);
+  EXPECT_TRUE(snapshot1 == snapshot2);
+
+  EXPECT_FALSE(snapshot1 != snapshot1);
+  EXPECT_FALSE(snapshot1 != snapshot2);
+}
+
+TEST_F(AggregateQuerySnapshotTest, Equality2) {
+  CollectionReference collection =
+      Collection({{"a", {{"k", FieldValue::String("a")}}},
+                  {"b", {{"k", FieldValue::String("b")}}},
+                  {"c", {{"k", FieldValue::String("c")}}}});
+  AggregateQuerySnapshot snapshot1 = ReadAggregate(collection.Count());
+  AggregateQuerySnapshot snapshot2 = ReadAggregate(collection.Count());
+
+  EXPECT_TRUE(snapshot1 == snapshot1);
+  EXPECT_TRUE(snapshot1 == snapshot2);
+
+  EXPECT_FALSE(snapshot1 != snapshot1);
+  EXPECT_FALSE(snapshot1 != snapshot2);
+}
+
+TEST_F(AggregateQuerySnapshotTest, Equality3) {
+  AggregateQuerySnapshot snapshot1 = AggregateQuerySnapshot();
+  AggregateQuerySnapshot snapshot2 = AggregateQuerySnapshot();
+
+  EXPECT_TRUE(snapshot1 == snapshot1);
+  EXPECT_TRUE(snapshot1 == snapshot2);
+
+  EXPECT_FALSE(snapshot1 != snapshot1);
+  EXPECT_FALSE(snapshot1 != snapshot2);
+}
+
+TEST_F(AggregateQuerySnapshotTest, NonEquality) {
+  CollectionReference collection =
+      Collection({{"a", {{"k", FieldValue::String("a")}}},
+                  {"b", {{"k", FieldValue::String("b")}}},
+                  {"c", {{"k", FieldValue::String("c")}}}});
+  AggregateQuerySnapshot snapshot1 = ReadAggregate(
+      collection.WhereEqualTo("k", FieldValue::String("d")).Count());
+  AggregateQuerySnapshot snapshot2 = ReadAggregate(collection.Limit(1).Count());
+  AggregateQuerySnapshot snapshot3 = ReadAggregate(collection.Limit(3).Count());
+  AggregateQuerySnapshot snapshot4 = ReadAggregate(collection.Count());
+  AggregateQuerySnapshot snapshot5 = AggregateQuerySnapshot();
+
+  EXPECT_TRUE(snapshot1 == snapshot1);
```

Comment:
Rather than hand-writing these equality checks, which is error prone and difficult to digest as a human reader, copy https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/test/unit/testutil/equals_tester.h from the ios sdk and use it. It makes equality and hash testing very easy. You just give it a bunch of "equality groups". Each "equality group" is a bunch of instances of the class that should compare equal. This also tests the Hash() function, as an added bonus.

Reply:
Ugh, looking at query_snapshot_test.cc, this appears to be how we test things. So, for consistency, you can just leave the test as it is currently written. Maybe add a TODO or TaskFlow task to replace it with the equality tester.